File: src/main/java/edu/univ/erp/api/admin/AdminApi.java
package edu.univ.erp.api.admin;

import edu.univ.erp.api.maintenance.MaintenanceApi;
import edu.univ.erp.data.AdminStore;
import edu.univ.erp.data.BackupManager;
import edu.univ.erp.data.SettingsStore;
import edu.univ.erp.domain.*;
import org.springframework.stereotype.Service;

import java.io.File;
import java.time.LocalDate;
import java.util.List;
import java.util.Map;

@Service
public class AdminApi {

    private final AdminStore adminStore;
    private final SettingsStore settingsStore;
    private final BackupManager backupManager;
    private final MaintenanceApi maintenanceApi;

    public AdminApi(AdminStore adminStore, SettingsStore settingsStore,
                    BackupManager backupManager, MaintenanceApi maintenanceApi) {
        this.adminStore = adminStore;
        this.settingsStore = settingsStore;
        this.backupManager = backupManager;
        this.maintenanceApi = maintenanceApi;
    }

    public List<UserAdminItem> getUsers() { return adminStore.getAllUsers(); }

    public String getDeadline() {
        LocalDate date = settingsStore.getRegistrationDeadline();
        return (date != null) ? date.toString() : "";
    }

    public void setDeadline(String date) { settingsStore.setRegistrationDeadline(date); }

    public boolean addUser(String username, String password, String role,
                           String extra1, String extra2, String extra3) {
        return adminStore.addUser(username, password, role, extra1, extra2, extra3);
    }

    public List<Course> getCourses() { return adminStore.getAllCourses(); }

    public boolean addCourse(String code, String title, int credits) {
        return adminStore.addCourse(code, title, credits);
    }

    public List<SectionAdminItem> getSections() { return adminStore.getAllSections(); }

    public Map<Integer, String> getInstructors() { return adminStore.getInstructorsMap(); }

    public boolean isMaintenanceOn() { return maintenanceApi.isMaintenanceOn(); }

    public void setMaintenance(boolean on) { maintenanceApi.setMaintenance(on); }

    public AdminStats getStats() { return adminStore.getDashboardStats(); }

    public boolean deleteUser(int userId) { return adminStore.deleteUser(userId); }

    public boolean deleteCourse(String code) { return adminStore.deleteCourse(code); }

    public boolean deleteSection(int sectionId) { return adminStore.deleteSection(sectionId); }

    public boolean updateSection(int sectionId, String dayTime, String room, int capacity) {
        return adminStore.updateSection(sectionId, dayTime, room, capacity);
    }

    public boolean addSection(String courseId, int instructorId, String dayTime, String room,
                              int capacity, String semester, int year) {
        return adminStore.addSection(courseId, instructorId, dayTime, room, capacity, semester, year);
    }

    public boolean toggleUserStatus(int userId, String currentStatus) {
        String newStatus = "Active".equalsIgnoreCase(currentStatus) ? "Inactive" : "Active";
        return adminStore.updateUserStatus(userId, newStatus);
    }

    public boolean triggerBackup(File file) { return backupManager.backup(file); }

    public boolean triggerRestore(File file) { return backupManager.restore(file); }
}

File: erp-web/lib/api/admin_api.dart
import 'client.dart';

class AdminApi {
  static Future<List<dynamic>> getUsers() async {
    final res = await apiClient.get('/admin/users');
    return res.data as List<dynamic>;
  }

  static Future<void> addUser(Map<String, String> data) async {
    await apiClient.post('/admin/users', data: data);
  }

  static Future<void> deleteUser(int userId) async {
    await apiClient.delete('/admin/users/$userId');
  }

  static Future<void> toggleUserStatus(int userId, String currentStatus) async {
    await apiClient.patch('/admin/users/$userId/status',
        data: {'currentStatus': currentStatus});
  }

  static Future<List<dynamic>> getCourses() async {
    final res = await apiClient.get('/admin/courses');
    return res.data as List<dynamic>;
  }

  static Future<void> addCourse(Map<String, dynamic> data) async {
    await apiClient.post('/admin/courses', data: data);
  }

  static Future<void> deleteCourse(String code) async {
    await apiClient.delete('/admin/courses/$code');
  }

  static Future<List<dynamic>> getSections() async {
    final res = await apiClient.get('/admin/sections');
    return res.data as List<dynamic>;
  }

  static Future<void> addSection(Map<String, dynamic> data) async {
    await apiClient.post('/admin/sections', data: data);
  }

  static Future<void> updateSection(
      int sectionId, Map<String, dynamic> data) async {
    await apiClient.patch('/admin/sections/$sectionId', data: data);
  }

  static Future<void> deleteSection(int sectionId) async {
    await apiClient.delete('/admin/sections/$sectionId');
  }

  static Future<List<dynamic>> getInstructors() async {
    final res = await apiClient.get('/admin/instructors');
    return res.data as List<dynamic>;
  }

  static Future<Map<String, dynamic>> getStats() async {
    final res = await apiClient.get('/admin/stats');
    return res.data as Map<String, dynamic>;
  }

  static Future<bool> getMaintenance() async {
    final res = await apiClient.get('/admin/maintenance');
    return res.data['maintenanceOn'] as bool;
  }

  static Future<void> setMaintenance(bool enabled) async {
    await apiClient.post('/admin/maintenance', data: {'enabled': enabled});
  }

  static Future<String> getDeadline() async {
    final res = await apiClient.get('/admin/deadline');
    return (res.data['deadline'] ?? '') as String;
  }

  static Future<void> setDeadline(String deadline) async {
    await apiClient.post('/admin/deadline', data: {'deadline': deadline});
  }

  static Future<void> backup(String path) async {
    await apiClient.post('/admin/backup', data: {'path': path});
  }

  static Future<void> restore(String path) async {
    await apiClient.post('/admin/restore', data: {'path': path});
  }
}
